src/setup_node_env/node_version_validator.js

```diff
@@ -28,27 +28,201 @@
-if (!isVersionValid) {
-  var errorMessage =
-    `OpenSearch Dashboards was built with ${requiredVersion} and does not support the current Node.js version ${currentVersion}. ` +
-    `Please use Node.js ${requiredVersion} or a higher patch version.`;
-
-  // Actions to apply when validation fails: error report + exit.
-  console.error(errorMessage);
-  process.exit(1);
+var pkgEngineNodeVersion = pkg && pkg.engines && pkg.engines.node;
+if (!pkgEngineNodeVersion || typeof pkgEngineNodeVersion !== 'string') {
+  console.error(
+    'OpenSearch Dashboards did not report its required version of the Node.js runtime. ' +
+      'Please revert any changes that might have been made to the package.json file and try again.'
+  );
+  process.exit(22);
+}
+
+/* Basic semver parsing: This is a very limited subset of what semver supports where only a single comparator, composed
+ * of an operator and a version, is supported.
+ * [https://github.com/npm/node-semver/blob/cb1ca1d5480a6c07c12ac31ba5f2071ed530c4ed/README.md#ranges]
+ *
+ * The supported operators are:
+ *   >    Greater than
+ *   >=   Greater than or equal to
+ *   =    Equal
+ *   ~    Tilde ranges: Allows patch changes if a minor version is specified but if only a major version is specified,
+ *          it allows minor changes.
+ *   ^    Caret ranges: Allows patch and minor updates when major is non-zero (and we will never have that).
+ *
+ * Note: If no operator is specified, equality is assumed.
+ */
+var requiredParts = pkgEngineNodeVersion.match(
+  /^\s*(>=?|\^|~|=)?\s*(\d+)(?:\.(\d+))?(?:\.(\d+))?(?:[^\d.].*)?$/
+);
+if (requiredParts === null) {
+  console.error(
+    'OpenSearch Dashboards did not report its required version of the Node.js runtime in a valid ' +
+      'format. Please revert any changes that might have been made to the package.json file and try again.'
+  );
+  process.exit(23);
+}
+
+var comparatorVersion = {
+  major: requiredParts[2],
+  minor: requiredParts[3],
+  patch: requiredParts[4],
+};
+var comparatorOperator = requiredParts[1] || '=';
+var rangeBottom = {
+  major: parseInt(comparatorVersion.major, 10) || 0,
+  minor: parseInt(comparatorVersion.minor, 10) || 0,
+  patch: parseInt(comparatorVersion.patch, 10) || 0,
+  inclusive: comparatorOperator.indexOf('>') === -1 || comparatorOperator.indexOf('=') > -1,
+};
+var rangeTop = undefined;
+
+if (!rangeBottom.major) {
+  console.error(
+    'OpenSearch Dashboards did not report its required version of the Node.js runtime in a valid ' +
+      'format. Please revert any changes that might have been made to the package.json file and try again.'
+  );
+  process.exit(24);
+}
+
+if (comparatorOperator === '>') {
+  if (!comparatorVersion.minor) {
+    // >3 is >=4.0.0
+    rangeBottom.major += 1;
+    rangeBottom.inclusive = true;
+  } else if (!comparatorVersion.patch) {
+    // >3.1 is >=3.2.0
+    rangeBottom.minor += 1;
+    rangeBottom.inclusive = true;
+  }
+}
```

Comment:
It seems like you could encapsulate `getRangeBottom` (L93-103) and `getRangeTop` (L110-141) into functions, which would be a little safer with the mutable values. But the layout as-is is organized well enough to follow.

Reply:
That was my only goal; if you have string feelings (which i dont) i will do it.

Reply:
Done.